File: backend/src/Domino.Backend/Application/Surveys/SurveysController.cs
using System.ComponentModel.DataAnnotations;
using System.Globalization;
using System.IdentityModel.Tokens.Jwt;
using Domino.Backend.Utilities;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Domino.Backend.Application.Surveys;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class SurveysController(ISurveysService service) : ControllerBase
{
    private int CurrentUserId =>
        int.Parse(User.FindFirst(JwtRegisteredClaimNames.Sub)!.Value, CultureInfo.InvariantCulture);

    [HttpGet("{slug}")]
    public async Task<IActionResult> GetBySlug(string slug)
    {
        var result = await service.GetBySlugAsync(slug);
        return ToActionResult(result);
    }

    [HttpPost("{slug}/responses")]
    public async Task<IActionResult> SubmitResponse(string slug, [FromBody] SubmitSurveyRequest request)
    {
        Console.WriteLine($"SubmitResponse: {slug}, {CurrentUserId}, {request.Answers}");
        if (!ModelState.IsValid)
        {
            return BadRequest(ModelState);
        }

        var result = await service.SubmitResponseAsync(slug, CurrentUserId, request.Answers);
        if (!result.IsSuccess)
        {
            return result.Error!.Kind switch
            {
                ServiceErrorKind.NotFound => NotFound(new { message = result.Error.Message }),
                _ => BadRequest(new { message = result.Error.Message }),
            };
        }

        return Ok(new { success = true });
    }

    [HttpGet("{slug}/responses/{userId:int}")]
    [Authorize(Roles = "Admin,SuperDuperAdmin")]
    public async Task<IActionResult> GetUserResponse(string slug, int userId)
    {
        var result = await service.GetUserResponseAsync(slug, userId);
        return ToActionResult(result);
    }

    private IActionResult ToActionResult<T>(ServiceResult<T> result)
    {
        if (result.IsSuccess)
        {
            return Ok(result.Value);
        }
        return result.Error!.Kind switch
        {
            ServiceErrorKind.NotFound => NotFound(new { message = result.Error.Message }),
            _ => BadRequest(new { message = result.Error.Message }),
        };
    }
}

public record SubmitSurveyRequest
{
    [Required]
    public required List<SubmitAnswerRequest> Answers { get; init; }
}

public record SubmitAnswerRequest
{
    [Required]
    public required int QuestionVersionId { get; init; }

    public string? TextValue { get; init; }
    public decimal? NumberValue { get; init; }
    public bool? BooleanValue { get; init; }
    public List<int>? SelectedOptionIds { get; init; }
}

File: backend-ts/src/surveys/surveys.controller.ts
import {
  Controller,
  Get,
  Post,
  Param,
  Body,
  ParseIntPipe,
  Res,
  UseGuards,
} from '@nestjs/common';
import type { Response } from 'express';
import { SurveysService } from './surveys.service';
import { SubmitSurveyRequest } from './dto/survey.dto';
import { JwtAuthGuard } from '../auth/guards/jwt-auth.guard';
import { RolesGuard } from '../auth/guards/roles.guard';
import { Roles } from '../auth/decorators/roles.decorator';
import { CurrentUser } from '../auth/decorators/current-user.decorator';
import type { AuthenticatedUser } from '../auth/decorators/current-user.decorator';
import { sendResult } from '../common/send-result';

@Controller('surveys')
@UseGuards(JwtAuthGuard)
export class SurveysController {
  constructor(private readonly service: SurveysService) {}

  @Get(':slug')
  async getBySlug(
    @Param('slug') slug: string,
    @Res({ passthrough: true }) res: Response,
  ) {
    const result = await this.service.getBySlug(slug);
    return sendResult(res, result);
  }

  @Post(':slug/responses')
  async submitResponse(
    @Param('slug') slug: string,
    @Body() request: SubmitSurveyRequest,
    @CurrentUser() currentUser: AuthenticatedUser,
    @Res({ passthrough: true }) res: Response,
  ) {
    const result = await this.service.submitResponse(
      slug,
      currentUser.userId,
      request.answers,
    );

    if (result.kind !== 'success') {
      return sendResult(res, result);
    }

    return { success: true };
  }

  @Get(':slug/responses/:userId')
  @UseGuards(RolesGuard)
  @Roles('Admin', 'SuperDuperAdmin')
  async getUserResponse(
    @Param('slug') slug: string,
    @Param('userId', ParseIntPipe) userId: number,
    @Res({ passthrough: true }) res: Response,
  ) {
    const result = await this.service.getUserResponse(slug, userId);
    return sendResult(res, result);
  }
}
